AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m

```diff
@@ -24,14 +54,206 @@ - (void)sendAsync:(NSURL *)url
+                                          event:^{
+                                            [self sendCallAsync:httpCall];
+                                          }];
+        return;
+      }
+
+      // Don't lose time pretty printing if not going to be printed.
+      if ([MSAppCenter logLevel] <= MSLogLevelVerbose) {
+        NSString *payload = [MSUtility prettyPrintJson:data];
+        MSLogVerbose([MSAppCenter logTag], @"HTTP response received with status code: %tu, payload:\n%@", httpResponse.statusCode, payload);
+      }
+    }
+    [self.pendingCalls removeObject:httpCall];
+  }
+
+  // Unblock the caller now with the outcome of the call.
+  httpCall.completionHandler(data, httpResponse, error);
+}
+
+- (void)networkStateChanged:(__unused NSNotificationCenter *)notification {
+  if ([self.reachability currentReachabilityStatus] == NotReachable) {
+    MSLogInfo([MSAppCenter logTag], @"Internet connection is down.");
+    [self pause];
+  } else {
+    MSLogInfo([MSAppCenter logTag], @"Internet connection is up.");
+    [self resume];
+  }
+}
+
+- (void)pause {
+  @synchronized(self) {
+    if (self.paused) {
+      return;
+    }
+    MSLogInfo([MSAppCenter logTag], @"Pause HTTP client.");
+    self.paused = YES;
+
+    // Reset retry for all calls.
+    for (MSHttpCall *call in self.pendingCalls) {
+      [call resetRetry];
+    }
+  }
+}
+
+- (void)resume {
+  @synchronized(self) {
+
+    // Resume only while enabled.
+    if (self.paused && self.enabled) {
+      MSLogInfo([MSAppCenter logTag], @"Resume HTTP client.");
+      self.paused = NO;
+
+      // Resume calls.
+      for (MSHttpCall *call in self.pendingCalls) {
+        if (!call.inProgress) {
+          [self sendCallAsync:call];
+        }
+      }
+    }
+  }
+}
+
+- (void)setEnabled:(BOOL)isEnabled {
+  @synchronized(self) {
+    if (self.enabled != isEnabled) {
+      _enabled = isEnabled;
+      if (isEnabled) {
+        self.session = [NSURLSession sessionWithConfiguration:self.sessionConfiguration];
+        [self.reachability startNotifier];
+        [self resume];
+      } else {
+        [self.reachability stopNotifier];
+        [self pause];
+
+        // Cancel all the tasks and invalidate current session to free resources.
+        [self.session invalidateAndCancel];
+        self.session = nil;
+
+        // Remove pending calls and invoke their completion handler.
+        for (MSHttpCall *call in self.pendingCalls) {
+          NSError *error = [NSError errorWithDomain:kMSACErrorDomain
+                                               code:kMSACCanceledErrorCode
+                                           userInfo:@{NSLocalizedDescriptionKey : kMSACCanceledErrorDesc}];
+          call.completionHandler(nil, nil, error);
+        }
+        [self.pendingCalls removeAllObjects];
+      }
+    }
+  }
+}
+
+- (NSString *)obfuscateHeaderValue:(NSString *)value forKey:(NSString *)key {
```

Comment:
Please also consider to refactor this one. Obfuscating app secret is only for Ingestion, not a good fit to stay in generic HttpClient.

Reply:
We actually still need that otherwise we can't log any headers in httpclient.

Reply:
My suggestion was synced offline. We can do this later since we only have appSecret that needs to be obfuscated.